File: scripts/lean-venue-catalog-images.mjs
/**
 * Fast-loading venue covers: catalog thumbs ~640px + hero cap ~1200px.
 *
 * `/images/*` bypasses `/_next/image` on MSK (nginx alias). Card grids must
 * not fetch 2-4MB GenerateImage originals. Writes sibling `-thumb.jpg` and
 * recompresses oversized originals in place.
 *
 *   node scripts/lean-venue-catalog-images.mjs
 */
import fs from 'node:fs';
import path from 'node:path';
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(__dirname, '..');
const publicVenues = path.join(rootDir, 'apps/public/public/images/venues');
const webVenues = path.join(rootDir, 'apps/web/public/images/venues');

const THUMB_WIDTH = 640;
const HERO_WIDTH = 1200;
const THUMB_QUALITY = 70;
const HERO_QUALITY = 76;
const LEAN_BYTES = 80 * 1024;
const HERO_SKIP_BYTES = 250 * 1024;
const CONCURRENCY = 3;

function loadSharp() {
  const candidates = [
    path.join(rootDir, 'apps/web/package.json'),
    path.join(rootDir, 'package.json'),
  ];
  let lastError;
  for (const pkg of candidates) {
    try {
      return createRequire(pkg)('sharp');
    } catch (error) {
      lastError = error;
    }
  }
  throw lastError || new Error('sharp not installed');
}

function walkJpgs(dir, acc = []) {
  if (!fs.existsSync(dir)) return acc;
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walkJpgs(full, acc);
      continue;
    }
    if (!/\.jpe?g$/i.test(entry.name)) continue;
    if (/-thumb\.jpe?g$/i.test(entry.name)) continue;
    acc.push(full);
  }
  return acc;
}

function thumbPathFor(file) {
  return file.replace(/\.jpe?g$/i, '-thumb.jpg');
}

function mirrorToWeb(absPublic) {
  if (!fs.existsSync(webVenues)) return;
  const rel = path.relative(publicVenues, absPublic);
  const dest = path.join(webVenues, rel);
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  fs.copyFileSync(absPublic, dest);
}

async function mapPool(items, limit, fn) {
  const stats = [];
  let index = 0;
  async function worker() {
    while (index < items.length) {
      const current = items[index];
      index += 1;
      stats.push(await fn(current));
    }
  }
  const n = Math.min(limit, items.length);
  await Promise.all(Array.from({ length: n }, () => worker()));
  return stats;
}

async function processFile(sharp, file) {
  const before = fs.statSync(file).size;
  const meta = await sharp(file).metadata();
  const width = Number(meta.width) || 0;
  const thumbFile = thumbPathFor(file);
  let originalAfter = before;
  let thumbBytes = 0;
  let wroteHero = false;
  let wroteThumb = false;

  const alreadyLean = before <= LEAN_BYTES && width > 0 && width <= THUMB_WIDTH;

  if (!alreadyLean && (width > HERO_WIDTH || before > HERO_SKIP_BYTES)) {
    const tmp = `${file}.hero-tmp.jpg`;
    await sharp(file)
      .rotate()
      .resize({ width: HERO_WIDTH, withoutEnlargement: true, fit: 'inside' })
      .jpeg({ quality: HERO_QUALITY, mozjpeg: true, progressive: true })
      .toFile(tmp);
    fs.renameSync(tmp, file);
    originalAfter = fs.statSync(file).size;
    wroteHero = true;
    mirrorToWeb(file);
  }

  const thumbExists = fs.existsSync(thumbFile);
  const thumbOk =
    thumbExists &&
    fs.statSync(thumbFile).size >= 8 * 1024 &&
    fs.statSync(thumbFile).size <= 120 * 1024;

  if (!thumbOk) {
    if (alreadyLean) {
      fs.copyFileSync(file, thumbFile);
    } else {
      await sharp(file)
        .rotate()
        .resize({ width: THUMB_WIDTH, withoutEnlargement: true, fit: 'inside' })
        .jpeg({ quality: THUMB_QUALITY, mozjpeg: true, progressive: true })
        .toFile(thumbFile);
    }
    wroteThumb = true;
    mirrorToWeb(thumbFile);
  }

  thumbBytes = fs.existsSync(thumbFile) ? fs.statSync(thumbFile).size : 0;
  originalAfter = fs.statSync(file).size;

  return {
    file: path.relative(publicVenues, file).replace(/\\/g, '/'),
    before,
    originalAfter,
    thumbBytes,
    wroteHero,
    wroteThumb,
  };
}

async function main() {
  const sharp = loadSharp();
  const files = walkJpgs(publicVenues);
  if (!files.length) {
    console.error('No venue jpgs under', publicVenues);
    process.exit(1);
  }

  const beforeTotal = files.reduce((sum, file) => sum + fs.statSync(file).size, 0);
  const results = await mapPool(files, CONCURRENCY, (file) => processFile(sharp, file));

  const originalAfter = results.reduce((sum, row) => sum + row.originalAfter, 0);
  const thumbTotal = results.reduce((sum, row) => sum + row.thumbBytes, 0);
  const thumbs = results.filter((row) => row.thumbBytes > 0);
  const typicalThumb = [...thumbs.map((row) => row.thumbBytes)].sort((a, b) => a - b)[
    Math.floor(thumbs.length / 2)
  ];
  const typicalBefore = [...results.map((row) => row.before)].sort((a, b) => a - b)[
    Math.floor(results.length / 2)
  ];

  console.log(
    JSON.stringify(
      {
        files: files.length,
        wroteHero: results.filter((row) => row.wroteHero).length,
        wroteThumb: results.filter((row) => row.wroteThumb).length,
        beforeMB: +(beforeTotal / 1e6).toFixed(1),
        originalAfterMB: +(originalAfter / 1e6).toFixed(1),
        thumbsMB: +(thumbTotal / 1e6).toFixed(1),
        typicalBeforeKB: Math.round((typicalBefore || 0) / 1024),
        typicalThumbKB: Math.round((typicalThumb || 0) / 1024),
      },
      null,
      2,
    ),
  );
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/lean-venue-catalog-images.py
"""
Fast-loading venue covers: catalog thumbs ~640px + hero cap ~1200px.

`/images/*` bypasses `/_next/image` on MSK (nginx alias). Card grids must
not fetch 2-4MB GenerateImage originals. Writes sibling `-thumb.jpg` and
recompresses oversized originals in place.

  python scripts/lean-venue-catalog-images.py
"""
from __future__ import annotations

import json
import shutil
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from PIL import Image, ImageOps

ROOT = Path(__file__).resolve().parents[1]
PUBLIC_VENUES = ROOT / "apps/public/public/images/venues"
WEB_VENUES = ROOT / "apps/web/public/images/venues"

THUMB_WIDTH = 640
HERO_WIDTH = 1200
THUMB_QUALITY = 70
HERO_QUALITY = 76
LEAN_BYTES = 80 * 1024
HERO_SKIP_BYTES = 250 * 1024
CONCURRENCY = 4


def walk_jpgs(directory: Path) -> list[Path]:
    if not directory.is_dir():
        return []
    files: list[Path] = []
    for path in directory.rglob("*"):
        if not path.is_file():
            continue
        name = path.name.lower()
        if not name.endswith((".jpg", ".jpeg")):
            continue
        if name.endswith("-thumb.jpg") or name.endswith("-thumb.jpeg"):
            continue
        files.append(path)
    return files


def thumb_path_for(file: Path) -> Path:
    return file.with_name(file.stem + "-thumb.jpg")


def mirror_to_web(abs_public: Path) -> None:
    if not WEB_VENUES.is_dir():
        return
    dest = WEB_VENUES / abs_public.relative_to(PUBLIC_VENUES)
    dest.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(abs_public, dest)


def save_jpeg(image: Image.Image, dest: Path, quality: int) -> None:
    rgb = image.convert("RGB")
    tmp = dest.with_name(dest.name + ".tmp.jpg")
    rgb.save(
        tmp,
        format="JPEG",
        quality=quality,
        optimize=True,
        progressive=True,
        subsampling=2,
    )
    tmp.replace(dest)


def fit_width(image: Image.Image, max_width: int) -> Image.Image:
    image = ImageOps.exif_transpose(image) or image
    if image.width <= max_width:
        return image
    ratio = max_width / float(image.width)
    height = max(1, int(image.height * ratio))
    return image.resize((max_width, height), Image.Resampling.LANCZOS)


def process_file(file: Path) -> dict:
    before = file.stat().st_size
    with Image.open(file) as opened:
        image = ImageOps.exif_transpose(opened) or opened
        image.load()
        width = image.width
        already_lean = before <= LEAN_BYTES and 0 < width <= THUMB_WIDTH
        wrote_hero = False

        working = image
        if not already_lean and (width > HERO_WIDTH or before > HERO_SKIP_BYTES):
            working = fit_width(image, HERO_WIDTH)
            save_jpeg(working, file, HERO_QUALITY)
            wrote_hero = True
            mirror_to_web(file)

        thumb_file = thumb_path_for(file)
        thumb_ok = (
            thumb_file.is_file()
            and 8 * 1024 <= thumb_file.stat().st_size <= 120 * 1024
        )
        wrote_thumb = False
        if not thumb_ok:
            if already_lean:
                shutil.copy2(file, thumb_file)
            else:
                source = working if wrote_hero else image
                save_jpeg(fit_width(source, THUMB_WIDTH), thumb_file, THUMB_QUALITY)
            wrote_thumb = True
            mirror_to_web(thumb_file)

        thumb_bytes = thumb_file.stat().st_size if thumb_file.is_file() else 0
        original_after = file.stat().st_size

    return {
        "file": file.relative_to(PUBLIC_VENUES).as_posix(),
        "before": before,
        "originalAfter": original_after,
        "thumbBytes": thumb_bytes,
        "wroteHero": wrote_hero,
        "wroteThumb": wrote_thumb,
    }


def median(values: list[int]) -> int:
    if not values:
        return 0
    ordered = sorted(values)
    return ordered[len(ordered) // 2]


def main() -> None:
    files = walk_jpgs(PUBLIC_VENUES)
    if not files:
        raise SystemExit(f"No venue jpgs under {PUBLIC_VENUES}")

    before_total = sum(path.stat().st_size for path in files)
    results: list[dict] = []
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as pool:
        futures = [pool.submit(process_file, path) for path in files]
        for future in as_completed(futures):
            results.append(future.result())

    original_after = sum(row["originalAfter"] for row in results)
    thumb_total = sum(row["thumbBytes"] for row in results)
    thumbs = [row["thumbBytes"] for row in results if row["thumbBytes"] > 0]
    print(
        json.dumps(
            {
                "files": len(files),
                "wroteHero": sum(1 for row in results if row["wroteHero"]),
                "wroteThumb": sum(1 for row in results if row["wroteThumb"]),
                "beforeMB": round(before_total / 1e6, 1),
                "originalAfterMB": round(original_after / 1e6, 1),
                "thumbsMB": round(thumb_total / 1e6, 1),
                "typicalBeforeKB": round(median([row["before"] for row in results]) / 1024),
                "typicalThumbKB": round(median(thumbs) / 1024),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()
